trl/trainer/grpo_trainer.py

```diff
@@ -390,6 +487,28 @@ def _get_per_token_logps(self, model, input_ids, attention_mask, logits_to_keep)
         return torch.stack(per_token_logps)
 
     def _prepare_inputs(self, inputs: dict[str, Union[torch.Tensor, Any]]) -> dict[str, Union[torch.Tensor, Any]]:
+        # When using vLLM, ne need to first update the model weights
+        if self.args.use_vllm and self.state.global_step != self._last_loaded_step:
+            with unwrap_model_for_generation(
+                self.model, self.accelerator, gather_deepspeed3_params=self.args.ds3_gather_for_generation
+            ) as unwrapped_model:
+                if is_compiled_module(unwrapped_model):
+                    state_dict = unwrapped_model._orig_mod.state_dict()
+                else:
+                    state_dict = unwrapped_model.state_dict()
+            if self.accelerator.is_main_process:
+                llm_model = self.llm.llm_engine.model_executor.driver_worker.model_runner.model
+                llm_model.load_weights(state_dict.items())
+            self._last_loaded_step = self.state.global_step
+
+        # Gather inputs and process them in the main process. This is important because the rewards are normalized
+        # per group.
+        inputs = gather_object(inputs)
+        prepared = self._prepare_main(inputs) if self.accelerator.is_main_process else None
```

Comment:
In the non-vllm setting, is only one GPU doing generation and the others sit idle?

Reply:
Yes, I hadn't thought of that.

Reply:
It shouldn't be super hard to solve that. Let's me try

Reply:
Done, it should be much easier to review now 🤗